### PR TITLE
Multiserver

### PR DESCRIPTION
**Pull request by jpelgrom**
## Summary

Add multiple servers to your app, send sensor updates to all of them and interact even when outside the app. Implements #296.

### Overview of changes

As changes to servers impact a lot of files, a small MVP wasn't really possible without breaking functionality. Here's an overview of the major changes to get started:

 - Introducing [`ServerManager`](https://github.com/jpelgrom/home-assistant-android/blob/multiserver/common/src/main/java/io/homeassistant/companion/android/common/data/servers/ServerManager.kt), which holds all information about servers known to the app and provides access to server-level repositories such as the `IntegrationRepository`.
   - There are two major types of [`Server`s](https://github.com/jpelgrom/home-assistant-android/blob/multiserver/common/src/main/java/io/homeassistant/companion/android/database/server/Server.kt): default and temporary servers. Temporary servers are currently used for onboarding and only exist in memory (= in the server manager, and are lost when removed or the app is destroyed).
   - The `AuthenticationRepository`, `IntegrationRepository` and `WebSocketRepository` now use assisted injection to get information about the server they should use and can no longer be injected directly.
   - The `UrlRepository` has been removed, most of the functionality can now be accessed directly using the server's [`ServerConnectionInfo`](https://github.com/jpelgrom/home-assistant-android/blob/multiserver/common/src/main/java/io/homeassistant/companion/android/database/server/ServerConnectionInfo.kt).
 - Once you add a server, when the server to use can't be inferred you'll see a new option appear to select one. This is usually pre-filled to the most recently used server in the app, and [automatically migrated for most existing settings/features](https://github.com/jpelgrom/home-assistant-android/blob/multiserver/common/src/main/java/io/homeassistant/companion/android/database/AppDatabase.kt#L640).
 - Sensors (state, attributes and settings) are currently shared across all servers. Enable it on one server, it's enabled for all. State updates are pushed to all servers.
   - In the database sensors get rows for each server to keep track of the last sent state and in case there are any differences in how data is handled due to the server version, but enabled and state is the same for all.
   - The only major exception to this rule is location tracking; while the enabled state is shared, it will take the individual server's zones into account when handling geofencing events and location tracking updates.

### Known limitations / out of scope

 - Wear OS doesn't expose any multiserver options; once you've added one server it works like the current version.
 - Switching servers is only available in the app settings, there are no gestures or panels yet.
 - App lock is a server-level feature and works as such, but isn't checked once the app settings are opened, allowing you to view/edit all server settings
 - Only a single TLS Client Certificate is supported.
 - Deeplinks (using `homeassistant://navigate` or `my.home-assistant.io`) will use the most recently used server instead of asking which server to use, as there is no server selection panel yet (see second limitation).

## Screenshots

Get started in app settings, where there's a new option "Add server":

|Light|Dark|
|-----|-----|
|App settings showing a section 'Servers & Devices', with one server, the option to add a new one, and Wear OS settings, light mode.|App settings showing a section 'Servers & Devices', with one server, the option to add a new one, and Wear OS settings, dark mode.|

After completing onboarding the server will be added. On the settings page, you can activate it, change the server name, or remove the server from the app.

|Light|Dark|
|-----|-----|
|A message at the bottom of the screen saying 'Server added' with a button labeled 'Activate', light mode.|A message at the bottom of the screen saying 'Server added' with a button labeled 'Activate', dark mode.|
|App settings showing a server with the name 'Home', and an option to activate the server, light mode.|App settings showing a server with the name 'Home', and an option to activate the server, dark mode.|
|Option 'Remove Server' at the bottom of the screen, light mode.|Option 'Remove Server' at the bottom of the screen, dark mode.|

If you receive a notification and have more than one server, the server name will be added to the notification:

|Light|Dark|
|-----|-----|
|A notification from the app that includes the server name 'Home' in the title, light mode.|A notification from the app that includes the server name 'Home' in the title, dark mode.|

And finally, some of the places where you'll see server choices if you have more than one:

|Light|Dark|
|-----|-----|
|Widget settings with a dropdown menu labeled 'Server', showing value 'Home', light mode.|Widget settings with a dropdown menu labeled 'Server', showing value 'Home', dark mode.|
|Shortcut configuration showing the option 'Select server' with value 'Home', light mode.|Shortcut configuration showing the option 'Select server' with value 'Home', dark mode.|
|Manage Device Controls settings screen with an option to switch entities for which server are shown, light mode.|Manage Device Controls settings screen with an option to switch entities for which server are shown, dark mode.|
|Sensor settings for 'High accuracy mode only when entering zone' showing a list with three zones from two different servers, light mode.|Sensor settings for 'High accuracy mode only when entering zone' showing a list with three zones from two different servers, dark mode.|

## Link to pull request in Documentation repository

Documentation: home-assistant/companion.home-assistant#910

## Any other notes

There are a lot of changes in this PR, so if you have any questions please feel free to ask here or on #devs_mobile_apps in Discord!

A lot of inspiration was taken from the Home Assistant iOS multiserver support and other open source Android apps with multi-user support.

I intend to add the following in followup PRs relatively soon after this PR is merged:
 - Add a gesture/panel to switch servers to match Home Assistant iOS
 - Check app lock when in settings and editing other servers
 - Allow enabling/disabling sensors per server

### CODE REVIEW

**Review by JBassett — APPROVED** [review]
I only found one thing, might just be that I burned out while reviewing the PR but I totally understand why the PR is so large.  I also testing it on my device for functionality and didn't notice any show stoppers getting this to beta.

File: app/src/main/java/io/homeassistant/companion/android/nfc/NfcViewModel.kt [hunk 1]
```diff
@@ -110,7 +110,7 @@ class NfcViewModel @Inject constructor(
         viewModelScope.launch {
             nfcTagIdentifier?.let {
                 try {
-                    integrationUseCase.scanTag(
+                    serverManager.integrationRepository().scanTag(
```
[the remark this answers is not in the document]
Comment (jpelgrom) [reply app/src/main/java/io/homeassistant/companion/android/nfc/NfcViewModel.kt:113]: Addressed in #3336

**Comment by mateuszdrab** [comment]
Is this already in the beta?
I can't wait to be able to try this out and feedback.
I have 5 HA instances I need to be able to flip between.
I used to use APK Cloner for this before but the app has become such a rip off I could almost say soon you would have to pay for it before you can open it 🫣

**Comment by dshokouhi** [comment]
> Is this already in the beta?

yes it is, check the releases section or sign up on the play store beta.

**Pull request merged**